<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
63   0.0067  0.0096  0.0024  0.0058  0.0197  0.0618  0.0432  0.0951  0.0836   
15   0.0298  0.0615  0.0650  0.0921  0.1615  0.2294  0.2176  0.2033  0.1459   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
146  0.1021  0.0830  0.0577  0.0627  0.0635  0.1328  0.0988  0.1787  0.1199   
153  0.0233  0.0394  0.0416  0.0547  0.0993  0.1515  0.1674  0.1513  0.1723   

         9   ...      51      52      53      54      55      56      57  \
63   0.1180  ...  0.0048  0.0023  0.0020  0.0040  0.0019  0.0034  0.0034   
15   0.0852  ...  0.0031  0.0153  0.0071  0.0212  0.0076  0.0152  0.0049   
164  0.2822  ...  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084  0.0108   
146  0.1369  ...  0.0709  0.0317  0.0309  0.0252  0.0087  0.0177  0.0214   
153  0.2078  ...  0.0104  0.0062  0.0026  0.0025  0.0061  0.0038  0.0101   

         58      59  60  
63   0.0051  0.0031   R  
15   0.0200  0.0073   R  
164  0.0083  0.0033   M  
146  0.0227  0.0106   M  
153  0.0078  0.0006   M  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60],
      dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

60
M    111
R     97
Name: count, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
36    True
48    True
66    True
168  False
81    True

In [10]:
y.value_counts()

R    
False    111
True      97
Name: count, dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [14]:
import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 755us/step - accuracy: 0.4665 - loss: 0.6933
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 620us/step - accuracy: 0.6281 - loss: 0.6793
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 608us/step - accuracy: 0.6385 - loss: 0.6687
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 567us/step - accuracy: 0.6913 - loss: 0.6497
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 514us/step - accuracy: 0.7543 - loss: 0.6150
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 422us/step - accuracy: 0.7422 - loss: 0.6092
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 405us/step - accuracy: 0.7422 - loss: 0.5460
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 433us/step - accuracy: 0.7873 - loss: 0.5404
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 401us/step - accuracy: 0.8422 - loss: 0.4908
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 438us/step - accuracy: 0.7941 - loss: 0.4791
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 443us/step - accuracy: 0.8366 - loss: 0.4312
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0

In [16]:
model.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7188 - loss: 0.8428 


[0.7403920292854309, 0.75]

Training Accuracy >>> Test Accuracy

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step
[3.5160679e-07 9.9645144e-01 6.9183356e-01 1.4172389e-05 9.9999994e-01
 9.9995905e-01 1.9100714e-01 9.9999988e-01 5.9004844e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  False
155  False
165  False
200  False
58    True
34    True
151  False
18    True
202  False
62    True

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.73      0.81      0.77        27
        True       0.77      0.68      0.72        25

    accuracy                           0.75        52
   macro avg       0.75      0.75      0.75        52
weighted avg       0.75      0.75      0.75        52



### Model with Dropout Layer

In [20]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 747us/step - accuracy: 0.5705 - loss: 0.6960
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 735us/step - accuracy: 0.5531 - loss: 0.7137
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 736us/step - accuracy: 0.5819 - loss: 0.6680
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 734us/step - accuracy: 0.5257 - loss: 0.6901
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 589us/step - accuracy: 0.5035 - loss: 0.6907
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 442us/step - accuracy: 0.5469 - loss: 0.6844
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 432us/step - accuracy: 0.5114 - loss: 0.7004
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 444us/step - accuracy: 0.5690 - loss: 0.6835
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 443us/step - accuracy: 0.4282 - loss: 0.7118
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 419us/step - accuracy: 0.5552 - loss: 0.6749
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 439us/step - accuracy: 0.5297 - loss: 0.6782
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0

In [21]:
modeld.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7837 - loss: 0.4608 


[0.450229287147522, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [22]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step
[6.3652021e-04 6.4048594e-01 9.3792230e-01 6.5088183e-02 9.9767995e-01
 9.4711542e-01 2.7398488e-01 9.9500954e-01 6.9568329e-02 9.9802804e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.76      0.81      0.79        27
        True       0.78      0.72      0.75        25

    accuracy                           0.77        52
   macro avg       0.77      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**